### PR TITLE
Render captcha modal with animation

### PR DESCRIPTION
**Pull request by pkong-ds**
## Preview - Light mode
https://github.com/user-attachments/assets/1df1e2c2-b1fa-4945-9993-85d7824fec57

## Preview - Dark mode
https://github.com/user-attachments/assets/a3200e75-b52f-4d7d-aef7-73d23d8d06c1

### CODE REVIEW

**Comment by pkong-ds** [comment]
@louischan-oursky ready for review, thanks!

**Comment by louischan-oursky** [comment]
@pkong-ds Please rebase.

**Comment by louischan-oursky** [comment]
@pkong-ds You need to run `check-tidy` locally.

File: resources/authgear/templates/en/web/authflowv2/__bot_protection_widget.html [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {{ if (eq $.BotProtectionProviderType "cloudflare" ) }}
-    <div class="flex justify-center" data-cloudflare-turnstile-target="widget"></div>
+    <div class="flex justify-center h-18" data-cloudflare-turnstile-target="widget"></div>
 {{ else if (eq $.BotProtectionProviderType "recaptchav2" ) }}
-    <div class="flex justify-center" data-recaptcha-v2-target="widget"></div>
+    <div class="flex justify-center h-19" data-recaptcha-v2-target="widget"></div>
```
Comment (louischan-oursky) [inline resources/authgear/templates/en/web/authflowv2/__bot_protection_widget.html:4]: Explain why.

File: authui/src/authflowv2/components/dialog.css [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+  }
+
+  @keyframes fadeOut {
+    0% {
+      display: flex;
+      opacity: 1;
+    }
+    99% {
+      opacity: 0.01;
+    }
```
Comment (louischan-oursky) [inline authui/src/authflowv2/components/dialog.css:31]: Explain why 99% is needed. Or you say `50% { opacity: 0.5; }`

File: authui/src/authflowv2/components/dialog.css [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+    --dialog-title__text-color: var(--color-neutral-700);
+    --dialog-description__text-color: var(--color-neutral-400);
+  }
+  :root.dark {
+    --dialog-title__text-color: var(--color-neutral-100);
+    --dialog-description__text-color: var(--color-neutral-200);
+  }
+  @keyframes fadeIn {
+    from {
+      display: flex;
```
Comment (louischan-oursky) [inline authui/src/authflowv2/components/dialog.css:16]: `display: flex;` seems redundant.

File: authui/src/authflowv2/components/dialog.css [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+  .dialog-title-description {
+    @apply flex flex-col gap-2;
+  }
+
+  .dialog-title {
+    font-family: inherit;
+    font-size: 1.5rem;
+    line-height: 2rem;
+    letter-spacing: 0;
+    font-weight: 700;
```
Comment (louischan-oursky) [inline authui/src/authflowv2/components/dialog.css:102]: Add these to base.css

```
    --typography-title-large__font-family: inherit;
    --typography-title-large__font-size: 1.5rem;
    --typography-title-large__line-height: 12em;
    --typography-title-large__letter-spacing: 0;
    --typography-title-large__font-weight: 700;
```

And then define

```
--dialog-title__font-family: var(--typography-title-large__font-family);
...
```

File: authui/src/authflowv2/components/dialog.css [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+      @apply text-end;
+    }
+  }
+
+  .dialog-description {
+    font-family: inherit;
+    font-size: 1rem;
+    line-height: 1.5rem;
+    letter-spacing: 0.03571428571em; /* 0.5 / 14 = 0.03571428571 */
+    font-weight: 500;
```
Comment (louischan-oursky) [inline authui/src/authflowv2/components/dialog.css:123]: Define 
```
--dialog-description__font-family: var(--typography-body-large__font-family);
...
```

File: authui/src/authflowv2/botprotection/botProtectionDialog.ts [hunk 6]
```diff
@@ -1,46 +1,54 @@
 import { Controller } from "@hotwired/stimulus";
-import { dispatchBotProtectionWidgetEventRender } from "./botProtectionWidget";
+import {
+  dispatchBotProtectionWidgetEventRender,
+  dispatchBotProtectionWidgetEventUndoRender,
+} from "./botProtectionWidget";
+
+const DIALOG_ID = "bot-protection-dialog";
+const BOT_PROTECTION_DIALOG_OPEN_EVENT = `dialog-${DIALOG_ID}:open`;
+const BOT_PROTECTION_DIALOG_CLOSE_EVENT = `dialog-${DIALOG_ID}:close`;
```
Comment (louischan-oursky) [inline authui/src/authflowv2/botprotection/botProtectionDialog.ts:9]: Remove line 8 and line 9.

File: authui/src/authflowv2/botprotection/botProtectionDialog.ts [hunk 7]
```diff
@@ -1,46 +1,54 @@
...
 
 /**
  * Dispatch a custom event to set captcha dialog open
  */
 export function dispatchBotProtectionDialogOpen() {
-  document.dispatchEvent(new CustomEvent("bot-protection-dialog:open"));
+  document.dispatchEvent(new CustomEvent(BOT_PROTECTION_DIALOG_OPEN_EVENT));
```
Comment (louischan-oursky) [inline authui/src/authflowv2/botprotection/botProtectionDialog.ts:15]: use dispatchDialogOpen here.

File: authui/src/authflowv2/botprotection/botProtectionDialog.ts [hunk 8]
```diff
@@ -1,46 +1,54 @@
...
 }
 
 /**
  * Dispatch a custom event to set captcha dialog close
  */
 export function dispatchBotProtectionDialogClose() {
-  document.dispatchEvent(new CustomEvent("bot-protection-dialog:close"));
+  document.dispatchEvent(new CustomEvent(BOT_PROTECTION_DIALOG_CLOSE_EVENT));
```
Comment (louischan-oursky) [inline authui/src/authflowv2/botprotection/botProtectionDialog.ts:22]: Use dispatchDialogClose here.

File: authui/src/authflowv2/dialog.ts [hunk 9]
```diff
@@ -28,8 +30,12 @@ export class DialogController extends Controller {
   };
 
   close = () => {
-    this.element.classList.remove("open");
     this.element.classList.add("close");
+    // We want close animation to show before `display: none` is applied
+    setTimeout(
+      () => this.element.classList.remove("open"),
+      CLOSE_ANIMATION_DURATION_MS
+    );
```
Comment (louischan-oursky) [inline authui/src/authflowv2/dialog.ts:38]: It does not seems a proper fix.

File: authui/src/authflowv2/components/close-btn.css [hunk 10]
```diff
@@ -0,0 +1,15 @@
+.close-btn {
+  @apply cursor-pointer;
+  
+  /* Make button clickable area at least 44px x 44px */
+  @apply p-5.5 -m-5.5;
+  transform: translateX(-0.2rem); /* FIXME: Figure out why button shifted right after applying above */
```
Comment (louischan-oursky) [inline authui/src/authflowv2/components/close-btn.css:6]: What?

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

**Comment by fungc-io** [comment]
Thanks, @pkong-ds @louischan-oursky 
The animation lgtm

File: authui/src/authflowv2/dialog.ts (hunk 9 above)
Comment (pkong-ds) [reply authui/src/authflowv2/dialog.ts]: I tried using `visibility: hidden -> visible` + `transformation` to replace existing `keyframe` + `animation`.

Seems achieved same result; ref [Use visibility over display for dialog animation](https://github.com/authgear/authgear-server/pull/4541/commits/1e69337b328d4ea8cc07ba39c4ce46b6b26f7bfa)

File: authui/src/authflowv2/components/close-btn.css (hunk 10 above)
Comment (pkong-ds) [reply authui/src/authflowv2/components/close-btn.css]: ref [this SO answer](https://stackoverflow.com/a/20327676/19287186)

Any better suggestion on `enlarging button clickable area`? Maybe `absolute` / `relative` position with `positive padding + negative margin`?

File: resources/authgear/templates/en/web/authflowv2/__bot_protection_widget.html [hunk 11]
```diff
@@ -1,5 +1,9 @@
+{{/* Note the fixed heights. This is to prevent the div having height=0 before widget render, which may cause UI glitching after render.
+    - h-16.5 = 66px > 65px = height of cloudflare turnstile widgets ref https://developers.cloudflare.com/turnstile/get-started/client-side-rendering/#widget-size
+    - h-19 = 76px = height of recaptchav2 widgets ref https://www.google.com/recaptcha/api2/demo
+*/}}
 {{ if (eq $.BotProtectionProviderType "cloudflare" ) }}
-    <div class="flex justify-center h-18" data-cloudflare-turnstile-target="widget"></div>
+    <div class="flex justify-center h-16.5" data-cloudflare-turnstile-target="widget"></div>
```
Comment (louischan-oursky) [inline resources/authgear/templates/en/web/authflowv2/__bot_protection_widget.html:6]: Then you should use `h-[65px]`.

File: authui/src/authflowv2/botprotection/botProtectionDialog.ts [hunk 12]
```diff
@@ -43,12 +42,12 @@ export class BotProtectionDialogController extends Controller {
       console.error(`bot-protection-dialog must have id="${DIALOG_ID}"`);
       return;
     }
-    document.addEventListener(BOT_PROTECTION_DIALOG_OPEN_EVENT, this.open);
-    document.addEventListener(BOT_PROTECTION_DIALOG_CLOSE_EVENT, this.close);
+    document.addEventListener(`dialog-${DIALOG_ID}:open`, this.open);
+    document.addEventListener(`dialog-${DIALOG_ID}:close`, this.close);
```
Comment (louischan-oursky) [inline authui/src/authflowv2/botprotection/botProtectionDialog.ts:46]: This still does not seem right.

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

File: authui/src/authflowv2/dialog.ts (hunk 9 above)
Comment (pkong-ds) [reply authui/src/authflowv2/dialog.ts]: okay, resolved

**Comment by pkong-ds** [comment]
@louischan-oursky updated, thanks!

https://github.com/user-attachments/assets/03787e21-22b7-4658-9fac-13275783347a

@fungc-io Note that after update, the captcha widget is rendered _after modal animation_. Please check if ok.

**Comment by pkong-ds** [comment]
@louischan-oursky updated, thanks!

https://github.com/user-attachments/assets/2eb874c1-1010-4cd5-bf4b-0ced49fe51f4

**Review by louischan-oursky — APPROVED** [review]

**Pull request merged**